### ISSUE TITLE
Limitations on current image generating method

### ISSUE THREAD

**Issue by lucasjinreal** [dvlab-research/MiniGemini#10]
Hi, I found currently the image generating way hard to make input image as reference:

![image](https://github.com/dvlab-research/MiniGemini/assets/21303438/e0c5a1bd-3ecf-44d7-a596-34465db1a141)

any thought about it?

BTW, I found the pretrain stage of this loss is very high:

```
{'loss': 2.573, 'learning_rate': 0.0007701925673852566, 'epoch': 0.34}                                                                                                                                                           
{'loss': 2.8217, 'learning_rate': 0.0007698799612970509, 'epoch': 0.34}                                                                                                                                                          
{'loss': 2.5646, 'learning_rate': 0.0007695672062744539, 'epoch': 0.34}                                                                                                                                                          
{'loss': 2.6857, 'learning_rate': 0.0007692543024900611, 'epoch': 0.34}                                                                                                                                                          
{'loss': 2.6944, 'learning_rate': 0.0007689412501165496, 'epoch': 0.34}                                                                                                                                                          
{'loss': 2.6418, 'learning_rate': 0.0007686280493266786, 'epoch': 0.34}                                                                                                                                                          
{'loss': 2.6801, 'learning_rate': 0.0007683147002932893, 'epoch': 0.34}                                                                                                                                                          
{'loss': 2.7245, 'learning_rate': 0.0007680012031893049, 'epoch': 0.34}                                                                                                                                                          
{'loss': 2.6275, 'learning_rate': 0.0007676875581877296, 'epoch': 0.34}
```

**Comment by yanwei-li** [comment]
Hi, because the generation process actually based on text in this version, it does not support fine-grained image editing. 
The loss seems high in epoch 0.34, it should be less than 2.0 in my experience. Could you please share your training script and model type?

**Comment by lucasjinreal** [comment]
@yanwei-li  This is my script:

```

AUX_SIZE=768


deepspeed train_xformers_gemini.py \
    --deepspeed ./scripts/gemini/zero2.json \
    --model_name_or_path ./checkpoints/$MODEL_VERSION \
    --version $PROMPT_VERSION \
    --data_path ./data/llava_0.1/pretrain_data.json \
    --image_folder ./data/images_all/pretrain_data \
    --vision_tower ./checkpoints/clip-vit-large-patch14-336 \
    --vision_tower_aux ./checkpoints/openclip-convnext-large-d-320-laion2B-s29B-b131K-ft-soup \
    --mm_projector_type mlp2x_gelu \
    --tune_mm_mlp_adapter True \
    --mm_vision_select_layer -2 \
    --mm_use_im_start_end False \
    --mm_use_im_patch_token False \
    --image_size_aux $AUX_SIZE \
    --bf16 False \
    --output_dir ./checkpoints/minigemini-$MODEL_VERSION-pretrain-siglip-ocr \
    --num_train_epochs 1 \
    --per_device_train_batch_size 2 \
    --per_device_eval_batch_size 4 \
    --gradient_accumulation_steps 8 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 3000 \
    --save_total_limit 1 \
    --learning_rate 1e-3 \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --tf32 False \
    --fp16 True \
    --model_max_length 2048 \
    --gradient_checkpointing True \
    --dataloader_num_workers 4 \
    --lazy_preprocess True

```

the training data just llava 1.5 pretrain data since I am finished data uncompressing of minigemini.

Am using V100x8 for pretrain.

**Comment by yanwei-li** [comment]
Hi, if you use original llava 1.5 training data, I guess loss > 2.0 is acceptable during pretraining. For Gemma, we actually tried to use plain instruction for pertaining, but the performance is much worse. I guess this could be attributed to the instruction-following abilities are quite different across LLMs, so we just use the same instruction format for both pertaining and sft for Gemma.

**Comment by lucasjinreal** [comment]
@yanwei-li Hi, why a template could effect final performance so much? Is there any deeper reason?

**Comment by yanwei-li** [comment]
Sorry for the incorrect expression, the performance difference is not too much but still has some effect. For example, in our experiments, when we change the stage-1 prompt to plain, the MME performance drops but the TextVQA increases a little. We maintain the original settings for convenience. 
It may be caused by the bias. I'm trying the plain version and will update the results here later.

**Comment by yanwei-li** [comment]
Hi, we align the data and change the prompt format in Stage-1 from `gemma`->`plain`, here is the result:

- Stage-1: `gemma`, Stage-2: `gemma`-> TextVQA: 56.2, MME: 1341/312, MM-Vet: 31.1
- Stage-1: `plain`, Stage-2: `gemma`-> TextVQA: 50.7, MME: 1318.6/283, MM-Vet: 29.4

Previously, we mainly conduct studies with relative small dataset. It's clear that when we improve the dataset to all instructions used in our approach, the gap will be further enlarged, especially with Gemma. This is not a general case for LLaMA-based LLMs. I guess this can be attributed to the Gemma-2b-it uses this format with most of data for their LLM training.

**Comment by lucasjinreal** [comment]
Oh, so that it could be better try same template in pretrain stage. Actually since the projector training stage didn't open LLM, it might be more suitable if keep same prompt as stage 2 since same projector were used.

**Comment by yanwei-li** [comment]
Theoretically, it is, but we do not find any significant gap in LLaMA-based LLMs.
